crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1388,84 +1392,35 @@ impl<'src> Lexer<'src> {
             return false;
         }
 
-        let mut current_position = self.current_range().start();
-        let mut reverse_chars = self.source[..current_position.to_usize()]
-            .chars()
-            .rev()
-            .peekable();
-        let mut newline_position = None;
-
-        while let Some(ch) = reverse_chars.next() {
-            if is_python_whitespace(ch) {
-                current_position -= ch.text_len();
-                continue;
-            }
-
-            match ch {
-                '\n' => {
-                    current_position -= ch.text_len();
-                    if let Some(carriage_return) = reverse_chars.next_if_eq(&'\r') {
-                        current_position -= carriage_return.text_len();
-                    }
-                }
-                '\r' => {
-                    current_position -= ch.text_len();
-                }
-                _ => break,
-            }
-
-            debug_assert!(matches!(ch, '\n' | '\r'));
-
-            // Count the number of backslashes before the newline character.
-            let mut backslash_count = 0;
-            while reverse_chars.next_if_eq(&'\\').is_some() {
-                backslash_count += 1;
-            }
+        let Some(new_position) = non_logical_newline_start else {
+            return false;
+        };
 
-            if backslash_count == 0 {
-                // No escapes: `\n`
-                newline_position = Some(current_position);
-            } else {
-                if backslash_count % 2 == 0 {
-                    // Even number of backslashes i.e., all backslashes cancel each other out
-                    // which means the newline character is not being escaped.
-                    newline_position = Some(current_position);
-                }
-                current_position -= TextSize::new('\\'.text_len().to_u32() * backslash_count);
-            }
+        // Earlier we reduced the nesting level unconditionally. Now that we know the lexer's
+        // position is going to be moved back, the lexer needs to be put back into a
+        // parenthesized context if the current token is a closing parenthesis.
+        //
+        // ```py
+        // (a, [b,
+        //     c
+        // )
+        // ```
+        //
+        // Here, the parser would request to re-lex the token when it's at `)` and can recover
+        // from an unclosed `[`. This method will move the lexer back to the newline character
+        // after `c` which means it goes back into parenthesized context.
+        if matches!(
+            self.current_kind,
```

Comment:
Does it still make sense to reduce the nesting level above unconditionally or could we invert the condition here and only then reduce the nesting?

Reply:
Yeah, good point, we can invert the condition

Reply:
Hmm, actually, it might still not be possible. Let me confirm